app/grandchallenge/discussion_forums/migrations/0002_migrate_old_forums.py

```diff
@@ -25,45 +186,76 @@ def migrate_challenge_forums(apps, schema_editor):
     topic_lock_matching_dict = {0: False, 1: True}
 
     for challenge in Challenge.objects.all():
-        new_forum = Forum.objects.create(
-            created=challenge.forum.created, modified=challenge.forum.updated
-        )
-        challenge.discussion_forum = new_forum
-        challenge.save()
+        print(f"Processing forum for {challenge.short_name}")
 
-        for topic in challenge.forum.topics.all():
-            new_topic = ForumTopic.objects.create(
-                forum=new_forum,
-                creator=topic.poster,
-                subject=topic.subject,
-                kind=topic_type_matching_dict[topic.type],
-                last_post_on=topic.last_post_on,
-                created=topic.created,
-                modified=topic.updated,
-                is_locked=topic_lock_matching_dict[topic.status],
+        try:
+            new_forum = challenge.forum.migrated_forum
+        except ObjectDoesNotExist:
+            new_forum = Forum.objects.create(
+                source_object=challenge.forum,
+                created=challenge.forum.created,
+                modified=challenge.forum.updated,
             )
+            challenge.discussion_forum = new_forum
+            challenge.save()
+            init_forum_permissions(apps=apps, challenge=challenge)
+
+        topic_count = 0
+        post_count = 0
+
+        for topic in challenge.forum.topics.all():
```

Comment:
Shouldn't ones that have already been migrated be excluded?

Reply:
No, we need to loop over all topics because a topic might have been migrated but there might be new posts for that topic.